segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -0,0 +1,382 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.val;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * An in-memory implementation of cache for read index that maps chunk start offset to chunk name for recently used segments.
+ * Eviction is performed only when number of segments or chunks exceeds certain given limits.
+ * Each time items are evicted the generation is incremented.
+ * For each entry in cache, we keep track of the generation in which it was last accessed.
+ * During eviction items with oldest generation are evicted first until enough objects are evicted.
+ * The least accessed segments are removed entirely before removing chunks from more recently used segments.
+ * This calculation is "best effort" and need not be accurate.
```

Comment:
Yes. But I think intention is to prefer performance over accuracy. 
Eg If it improves average performance by avoiding need to synchronize we would prefer that option, even if every once in a while we evict an entry too early or keep it for bit longer than needed.

It really depends on extent of miscalculation and probability of it happening.

Reply:
I need to understand this better, I'll have a look at this again and get back.

Reply:
Ok, I'm trying to understand what made you say best effort here. Is it the fact that upon eviction, we iterate over the elements of the map concurrently with other operations? If not, what's it?

I also saw a resolved comment down saying that you'll revisit the implementation of the cache in the next iteration? It is not clear to me what the next iteration is, whether for this PR or a future one, so let me know your plans so that I have the right expectation.

Reply:
Resolving as per discussion offline. Tracking this here https://github.com/pravega/pravega/issues/4902